dbos/_core/system_database.py

```diff
@@ -0,0 +1,1734 @@
+            cmd = cmd.on_conflict_do_update(
+                index_elements=["workflow_uuid"],
+                set_=dict(
+                    recovery_attempts=SystemSchema.workflow_status.c.recovery_attempts
+                    + 1,
+                ),
+            )
+        else:
+            cmd = cmd.on_conflict_do_nothing()
+        return cmd.returning(SystemSchema.workflow_status.c.recovery_attempts)
+
+    def update_workflow_status_sync(
+        self,
+        status: WorkflowStatusInternal,
+        replace: bool = True,
+        in_recovery: bool = False,
+        *,
+        conn: Optional[sa.Connection] = None,
+        max_recovery_attempts: int = DEFAULT_MAX_RECOVERY_ATTEMPTS,
+    ) -> None:
+        cmd = self._get_update_workflow_status_cmd(status, replace, in_recovery)
+
+        if conn is not None:
+            results = conn.execute(cmd)
+        else:
+            with self.sync_engine.begin() as c:
+                results = c.execute(cmd)
+
+        if in_recovery:
+            row = results.fetchone()
+            if row is not None:
+                recovery_attempts: int = row[0]
+                if recovery_attempts > max_recovery_attempts:
+                    with self.sync_engine.begin() as c:
+                        c.execute(
+                            sa.update(SystemSchema.workflow_status)
+                            .where(
+                                SystemSchema.workflow_status.c.workflow_uuid
+                                == status["workflow_uuid"]
+                            )
+                            .where(
+                                SystemSchema.workflow_status.c.status
+                                == WorkflowStatusString.PENDING.value
+                            )
+                            .values(
+                                status=WorkflowStatusString.RETRIES_EXCEEDED.value,
+                            )
+                        )
+                    raise DBOSDeadLetterQueueError(
+                        status["workflow_uuid"], max_recovery_attempts
+                    )
+
+        # Record we have exported status for this single-transaction workflow
+        if status["workflow_uuid"] in self._temp_txn_wf_ids:
+            self._exported_temp_txn_wf_status.add(status["workflow_uuid"])
+
+    async def update_workflow_status_async(
+        self,
+        status: WorkflowStatusInternal,
+        replace: bool = True,
+        in_recovery: bool = False,
+        *,
+        conn: Optional[AsyncConnection] = None,
+        max_recovery_attempts: int = DEFAULT_MAX_RECOVERY_ATTEMPTS,
+    ) -> None:
+        cmd = self._get_update_workflow_status_cmd(status, replace, in_recovery)
+
+        if conn is not None:
+            results = await conn.execute(cmd)
+        else:
+            async with self.async_engine.begin() as c:
+                results = await c.execute(cmd)
+        if in_recovery:
+            row = results.fetchone()
```

Comment:
I am surprised this is not async.  🤷

Reply:
The await occurs above at the `execute` call. I suspect this is SQLAlchemy's async wrapping strategy peeking thru.